File: LeetCode/Problems/C++/173. Binary Search Tree Iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
private:
    stack<TreeNode*> stack;
    TreeNode* head;
public:
    BSTIterator(TreeNode* root) {
        head = root;
        TreeNode* node = head;
        while(node!=NULL){
            stack.push(node);
            node = node->left;
        }
    }
    void checkRightSubTree(TreeNode* node) {
        if(node->right){
            stack.push(node->right);
            while(stack.top()->left != NULL)
                stack.push(stack.top()->left);
        } 
    }
    /** @return the next smallest number */
    int next() {
        TreeNode* curr = stack.top();
        stack.pop();
        checkRightSubTree(curr);
        return curr->val;
    }
    
    /** @return whether we have a next smallest number */
    bool hasNext() {
        return !stack.empty();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: LeetCode/Problems/Python/173. Binary Search Tree Iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: TreeNode):
        self.root = root
        self.stack = []
        node = root
        while node:
            self.stack.append(node)
            node = node.left

    def next(self) -> int:
        """
        @return the next smallest number
        """
        self.hasNext()
        curr = self.stack.pop()
        if curr.right:
            self.stack.append(curr.right)
            while self.stack[-1].left:
                self.stack.append(self.stack[-1].left)
        
        return curr.val
        

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return len(self.stack) > 0
        

# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()